src/ancestry_proof.rs

```diff
@@ -0,0 +1,342 @@
+                    return Err(Error::CorruptedProof);
+                };
+                (parent_pos, parent_item)
+            }
+        };
+
+        if parent_pos <= peak_pos {
+            let parent = (parent_pos, parent_item, height + 1);
+            if peak_pos == parent_pos
+                || queue.front() != Some(&parent)
+                    && !sibs_processed_from_back.iter().any(|item| item == &parent)
+            {
+                queue.push_front(parent)
+            };
+        } else {
+            return Err(Error::CorruptedProof);
+        }
+    }
+    Err(Error::CorruptedProof)
+}
+
+fn calculate_peaks_hashes<
+    'a,
+    T: 'a + PartialEq + Clone,
+    M: Merge<Item = T>,
+    I: Iterator<Item = &'a (u64, T)>,
+>(
+    nodes: Vec<(u64, T)>,
+    mmr_size: u64,
+    proof_iter: I,
+) -> Result<Vec<T>> {
+    // special handle the only 1 leaf MMR
+    if mmr_size == 1 && nodes.len() == 1 && nodes[0].0 == 0 {
+        return Ok(nodes.into_iter().map(|(_pos, item)| item).collect());
+    }
+
+    // ensure nodes are sorted and unique
+    let mut nodes: Vec<_> = nodes
+        .into_iter()
+        .chain(proof_iter.cloned())
+        .sorted_by_key(|(pos, _)| *pos)
+        .dedup_by(|a, b| a.0 == b.0)
+        .collect();
+
+    let peaks = get_peaks(mmr_size);
+
+    let mut peaks_hashes: Vec<T> = Vec::with_capacity(peaks.len() + 1);
+    for peak_pos in peaks {
+        let mut nodes: Vec<(u64, T)> = take_while_vec(&mut nodes, |(pos, _)| *pos <= peak_pos);
+        let peak_root = if nodes.len() == 1 && nodes[0].0 == peak_pos {
+            // leaf is the peak
+            nodes.remove(0).1
+        } else if nodes.is_empty() {
+            // if empty, means the next proof is a peak root or rhs bagged root
+            // means that either all right peaks are bagged, or proof is corrupted
+            // so we break loop and check no items left
+            break;
+        } else {
+            calculate_peak_root::<_, M>(nodes, peak_pos)?
+        };
+        peaks_hashes.push(peak_root.clone());
+    }
+
+    // ensure nothing left in leaves
+    if nodes.len() != 0 {
+        return Err(Error::CorruptedProof);
+    }
+
+    // check rhs peaks
+    // if let Some((_, rhs_peaks_hashes)) = proof_iter.next() {
+    //     peaks_hashes.push(rhs_peaks_hashes.clone());
+    // }
+    // ensure nothing left in proof_iter
+    // if proof_iter.next().is_some() {
+    //     return Err(Error::CorruptedProof);
+    // }
+    Ok(peaks_hashes)
+}
+
+pub fn bagging_peaks_hashes<T, M: Merge<Item = T>>(mut peaks_hashes: Vec<T>) -> Result<T> {
+    // bagging peaks
+    // bagging from right to left via hash(right, left).
+    while peaks_hashes.len() > 1 {
+        let right_peak = peaks_hashes.pop().expect("pop");
+        let left_peak = peaks_hashes.pop().expect("pop");
+        peaks_hashes.push(M::merge_peaks(&right_peak, &left_peak)?);
+    }
+    peaks_hashes.pop().ok_or(Error::CorruptedProof)
+}
```

Comment:
Nit: I would just use `crate::mmr::bagging_peaks_hashes()` instead of duplicating it.